src/ir_generator.rs

```diff
@@ -0,0 +1,193 @@
+            instructions: Default::default(),
+            register_count: 0,
+            file_count: 0,
+        }
+    }
+
+    /// Generates the IR from the given state of the compiler.
+    /// After this is called, use `block` and `errors` to get the result.
+    pub fn generate(&mut self) {
+        if self.compiler.ast_nodes.is_empty() {
+            return;
+        }
+        let Some(reg) = self.generate_node(NodeId(self.compiler.ast_nodes.len() - 1)) else {
+            return;
+        };
+        self.instructions.push(Instruction::Return { src: reg });
+    }
+
+    /// Returns generated IR block.
+    ///
+    /// Call `generate` before using this method and ensure there are no errors.
+    pub fn block(self) -> IrBlock {
+        // TODO: properly generate the spans
+        // TODO: figure out what to do with AST, as this parser has different
+        // representation of AST than the old parser.
+        let mut spans = vec![];
+        let mut ast = vec![];
+        for _ in 0..(self.instructions.len()) {
+            spans.push(Span { start: 0, end: 0 });
+            ast.push(None);
+        }
+        IrBlock {
+            instructions: self.instructions,
+            spans,
+            data: Default::default(),
+            ast,
+            comments: Default::default(),
+            register_count: self.register_count,
+            file_count: self.file_count,
+        }
+    }
+
+    /// Returns errors encountered during IR generation step.
+    ///
+    /// Call `generate` before using this method.
+    pub fn errors(&self) -> &Vec<SourceError> {
+        &self.errors
+    }
+
+    /// Prints the internal state to standard output.
+    pub fn print(&self) {
+        let output = self.display_state();
+        print!("{output}");
+    }
+
+    /// Displays the state of the IR generator.
+    /// The output can be used for human debugging and for snapshot tests.
+    pub fn display_state(&self) -> String {
+        let mut result = String::new();
+        result.push_str("==== IR ====\n");
+        result.push_str(&format!("register_count: {}\n", self.register_count));
+        result.push_str(&format!("file_count: {}\n", self.file_count));
+
+        for (idx, instruction) in self.instructions.iter().enumerate() {
+            result.push_str(&format!("{}: {:?}\n", idx, instruction));
+        }
+
+        if !self.errors.is_empty() {
+            result.push_str("==== IR ERRORS ====\n");
+            for error in &self.errors {
+                result.push_str(&format!(
+                    "{:?} (NodeId {}): {}\n",
+                    error.severity, error.node_id.0, error.message
+                ));
+            }
+        }
+        result
+    }
+
+    // Returns unused register.
+    fn next_register(&mut self) -> RegId {
+        let r = RegId::new(self.register_count);
+        self.register_count += 1;
+        r
+    }
+
+    fn span_to_string(&mut self, node_id: NodeId) -> Option<String> {
```

Comment:
I can this use [std::str::from_utf8()](https://doc.rust-lang.org/std/str/fn.from_utf8.html) instead and return `&str` to avoid the String allocation?

Also, I'd move these two `span_to_xxx()` functions to the Compiler, and perhaps rename then to `node_as_string/i64()`, or similar.